src/Generators/Microsoft.Gen.Logging/Emission/Emitter.Method.cs

```diff
@@ -276,132 +165,317 @@ static string ConvertPropertyToString(LoggingProperty lp, string arg)
+                        else
+                        {
+                            var value = ShouldStringify(p.Type)
+                                ? ConvertToString(p, p.NameWithAt)
+                                : p.NameWithAt;
+
+                            OutLn($"{stateName}.PropertyArray[{--count}] = new({key}, {value});");
+                        }
                     }
-                    else
+                }
+
+                foreach (var p in lm.Parameters)
+                {
+                    if (p.HasProperties && !p.SkipNullProperties)
                     {
-                        OutLn($"var {atSign}{t.Key} = {lambdaHelperName}[{holderMap[parameter]}].Value;");
-                        result = true;
+                        p.TraverseParameterPropertiesTransitively((propertyChain, member) =>
+                        {
+                            if (member.ClassificationAttributeType == null)
+                            {
+                                var propName = PropertyChainToString(propertyChain, member, "_", omitParameterName: p.OmitParameterName);
+                                var accessExpression = PropertyChainToString(propertyChain, member, "?.", nonNullSeparator: ".");
+
+                                var ts = ShouldStringify(member.Type)
+                                    ? ConvertPropertyToString(member, accessExpression)
+                                    : accessExpression;
+
+                                var value = member.IsEnumerable
+                                    ? $"{LoggerMessageHelperType}.Stringify({accessExpression})"
+                                    : ts;
+
+                                OutLn($"{stateName}.PropertyArray[{--count}] = new(\"{propName}\", {value});");
+                            }
+                        });
                     }
                 }
-            }
-
-            return result;
-        }
 
-        static (string name, bool isNullable) GetLogger(LoggingMethod lm)
-        {
-            string logger = lm.LoggerField;
-            bool isNullable = lm.LoggerFieldNullable;
+                if (!string.IsNullOrEmpty(lm.Message))
+                {
+                    OutLn($"{stateName}.PropertyArray[{--count}] = new(\"{{OriginalFormat}}\", \"{EscapeMessageString(lm.Message)}\");");
+                }
+            }
 
-            foreach (var p in lm.Parameters)
+            if (numReservedClassifiedProps > 0)
             {
-                if (p.IsLogger)
+                OutLn();
+                OutLn($"_ = {stateName}.ReserveClassifiedPropertySpace({numReservedClassifiedProps});");
+                int count = numReservedClassifiedProps;
+                foreach (var p in lm.Parameters)
                 {
-                    logger = p.Name;
-                    isNullable = p.IsNullable;
-                    break;
+                    if (NeedsASlot(p) && p.ClassificationAttributeType != null)
```

Comment:
Can the check `ClassificationAttributeType != null` be extracted into a get-only property on the type itself?

Reply:
Done.